src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java

```diff
@@ -262,38 +261,56 @@ SecurityAnalysisResult runSimulationsOnAllComponents(LfNetworkList networks, Lis
-            for (LfNetwork n : networkToSimulate) {
-                SecurityAnalysisResult resultOtherComponent = runSimulations(n, propagatedContingencies, parameters, securityAnalysisParameters, operatorStrategies, actions, limitReductions);
+        List<LfNetwork> networkToSimulate = selectValidSecondaryComponents(networks, lfParameters.getConnectedComponentMode());
 
-                // Merge into first result
-                // PreContingency results first
-                preContingenctViolations.addAll(resultOtherComponent.getPreContingencyResult().getLimitViolationsResult().getLimitViolations());
-                mergedPreContingencyNetworkResult = mergeNetworkResult(mergedPreContingencyNetworkResult, resultOtherComponent.getPreContingencyResult().getNetworkResult());
+        for (LfNetwork n : networkToSimulate) {
+            SecurityAnalysisResult resultOtherComponent = runSimulations(n, propagatedContingencies, parameters, securityAnalysisParameters, operatorStrategies, actions, limitReductions);
 
-                // PostContingency and OperatorStrategies results
-                mergeSecurityAnalysisResult(resultOtherComponent, postContingencyResultMap, operatorStrategyResultMap, n.getNumCC());
-            }
-            postContingencyResults = postContingencyResultMap.values().stream().toList();
-            operatorStrategyResults = operatorStrategyResultMap.values().stream().toList();
+            // Merge into first result
+            // PreContingency results first
+            preContingenctViolations.addAll(resultOtherComponent.getPreContingencyResult().getLimitViolationsResult().getLimitViolations());
+            mergedPreContingencyNetworkResult = mergeNetworkResult(mergedPreContingencyNetworkResult, resultOtherComponent.getPreContingencyResult().getNetworkResult());
+
+            // PostContingency and OperatorStrategies results
+            mergeSecurityAnalysisResult(resultOtherComponent, postContingencyResultMap, operatorStrategyResultMap, n.getNumCC());
         }
+        postContingencyResults = postContingencyResultMap.values().stream().toList();
+        operatorStrategyResults = operatorStrategyResultMap.values().stream().toList();
+
         PreContingencyResult mergedPrecontingencyResult =
                 new PreContingencyResult(result.getPreContingencyResult().getStatus(),
                         new LimitViolationsResult(preContingenctViolations),
                         mergedPreContingencyNetworkResult);
         return new SecurityAnalysisResult(mergedPrecontingencyResult, postContingencyResults, operatorStrategyResults);
     }
 
+    static Optional<LfNetwork> selectValidMainComponent(LfNetworkList networks) {
+        return networks.getList().stream().filter(n -> isMainCCandMainSC(n) && n.getValidity().equals(LfNetwork.Validity.VALID)).findFirst();
+    }
```

Comment:
method name `selectValidMainComponent` is a confusing since we don't know if this means main CC or main SC.

It would be clearer with one method `List<LfNetwork> getNetworksToSimulate(LoadFlowParameters.ConnectedComponentMode mode)`

- in mode MAIN: returns all valid CC0 SCx
- in mode ALL: retuens all valid CCx SCy

LfNetworkList would be the right place for this method.

Reply:
That is not what needs to be done here. I understand your approach but : 
We need a way here to find a "single" main component (usually CC0|SC0) to initiate the result structure. Here your MAIN mode would yield a list of several component on which we should again select the first one.

In fact thos methods are highly linked to the security analysis flow and only are this way because of the constraints on merged results building. 
This is also why I kept that in AbstractSecurityAnalysis, this is not supposed to be used elsewhere and will probably disappear when we rethink a more suitable structure for merged result someday.

Reply:
I think this is exactly what needs to be done. Expectation is:
- in mode MAIN: runs all valid CC0 SCx
- in mode ALL: runs all valid CCx SCy
But the code is much more complicated, only to accommodate technical issues.
Also there is the assumption that CC0 SC0 always exists, which is as you wrote "usually" true / "usually" works, but can be wrong, I added a test case to demonstrate this.
Could be done later on if you prefer, but I find the code hard to follow.